### PR TITLE
allow positional access in named structs

### PR DESCRIPTION
**Pull request by peterboncz**
allow positional access in named structs

the reason why I propose this is not only to allow more flexible access to structs, but I also would like to fix an issue in TableFilter:ToExpression() which in case of predicates on struct members fails to produce a working boolean expression (because, it creates a positional access rather than a key-based one).

It actually also got that position wrong (off by one).

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR!

Perhaps we can have a different method that does this? Overloading the struct extract means it becomes possible to index structs like arrays (i.e. `struct_col[1]`) which I think is undesirable/potentially confusing.

**Comment by peterboncz** [comment]
Right, something like struct_extract_by_position() ?

Two remarks though:
- I suppose st[key] is syntactic sugar for struct_extract(st, key), so defining a new method would mean that the syntactic sugar won't work for the positional access and there is no amboiguity
- However, st[1] that is positional access is already supported now. But it is only supported on structs that have no names for the keys (isUnnamed()) - so would it not be just a matter of consistency to allow it?

**Comment by Mytherin** [comment]
Unnamed structs are not really structs, they are more like tuples in Python. They don't have names so elements must be accessed by position. The existing struct extract syntax can keep working for them as it works in python, e.g.:

```py
>>> a = (42, 84);
>>> a[1]
84
>>> a = {'key1': 42, 'key2': 84}
>>> a[1]
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    a[1]
    ~^^^
KeyError: 1

```

**Comment by peterboncz** [comment]
went for the name struct_extract_at(st, pos) because it is shorter

File: src/function/scalar/struct/struct_extract.cpp [hunk 1]
```diff
@@ -87,8 +87,8 @@ static unique_ptr<FunctionData> StructExtractBind(ClientContext &context, Scalar
 	return GetBindData(key_index);
 }
 
-static unique_ptr<FunctionData> StructExtractBindIndex(ClientContext &context, ScalarFunction &bound_function,
-                                                       vector<unique_ptr<Expression>> &arguments) {
+static unique_ptr<FunctionData> StructExtractBindInternal(ClientContext &context, ScalarFunction &bound_function,
+                                                          vector<unique_ptr<Expression>> &arguments, bool by_pos) {
```
Comment (Mytherin) [inline src/function/scalar/struct/struct_extract.cpp:91]: Minor nit, but this bool should probably be `bool by_name`

**Review by Mytherin — COMMENTED** [review]
Thanks! LGTM - one minor nit

File: src/function/scalar/struct/struct_extract.cpp (hunk 1 above)
Comment (peterboncz) [reply src/function/scalar/struct/struct_extract.cpp]: both functions actually do 'by_pos'. 

I renamed the bool 'struct_extract' now to distinguish between the behavior of 'struct_extract()' and 'struct_extract_at()' => the former only allows positional access on unnamed structs.

**Pull request merged**